### ISSUE TITLE
Wish list: Queue building placement.

### ISSUE THREAD

**Issue by IkkeTM** [OpenRA/OpenRA#15847]
The feature request, in short, is for one to be able to place "ghosts" or placeholders of a building when one adds it to the queue, and having it placed there automatically once the building is ready. The queue itself would remain unchanged aside from this automatic placement.

**Mechanics:**
Here is how it should work:

When one orders a building, a ghost of the building will appear under the mouse cursor when it is on the game map. Left click places the ghost, right click dismisses placing the ghost (without canceling production).
If multiple buildings are ordered in succession, the first building ordered will be the first under the cursor, and when placed or dismissed, the next one will appear, and etc. 

Ghosts can not be placed on top of each other, existing structures or illegal terrain. They are however not influenced by building radius or adjacency. Ghosts can be moved again by clicking them with the left mouse button, if when clicked upon, there is no other selectable object where the mouse cursor is.
If a building is removed from the building queue, its ghost is also removed.

If a building cannot be placed where its ghost is, for whatever reason, maybe a unit is in the way, maybe the MCV was destroyed, undeployed or not yet deployed, maybe the adjacency was misjudged, 3 things happen.
1) A warning (audio) message is sent to the player “building placement blocked”.
2) The game will recheck if it can place the building at the desired location at reasonable intervals.
3) Until the building is placed, the placement defaults to manual placement, including a lit building icon to remind the player what's up.

Edit: For balancing purposes, it might be neccesary to add a delay factor between finishing a building's production and it getting placed on the map at the ghost's location, as a manual player will take some time to do this. Something in the order of 2 to 5 seconds.

**Rationale:**
The good thing about the way CnC's queue's work, if you compare them to those of other RTS's, in almost every other game you have to find one specific building where the unit is built, order them up there, then do the same for the next building and etc. Then rally them all. This is a very labor intensive way of ordering up units. 

CnC's production tabs are very good at mitigating this. You open the tab, you order the units you want, you never leave where your attention was in the first place, and eventually pick up your units by the rally point, which is just and about the only thing you need to set by moving your camera back where the production facility is. This is a great design decision, you're not dealing with the buildings, you order up what you want and you get it delivered. You just focus on the fun stuff.

The one thing for which this UI/design decision works out less well, is the building queue. Building placement is an unwieldy two-moment affair of ordering it up, waiting for it to be done, then find a place to put it. Then repeat this piecemeal every 15-40 second for every next building and so on. Thus each individual building takes up a few seconds of time, which over the course of a game can to add up (5-10% of game time), even if you use [H]-[E]-[leftclick]-[1][1] (or another hotkey group to get back to where you were). Everyone messes this up from time to time.

Now when you look at the original Westwood maps, or the campaign missions, the ore patches were almost always such that you play one base, and maybe, sometimes, an expansion. In the original designer's intent, there was no constant base building, your base is done after 10-15 buildings, and at any rate is in one location. It's in the way that OpenRA designs maps and plays the game in multi, with multiple expansions and all, that the building queue keeps running indefinitely and thus interrupts the otherwise well tailored "minimal hassle" production design.

Perhaps then OpenRA can take a page from those RTSs where base planning is concerned, as nearly all those games allow one to place several buildings in a row, either by setting workers on a waypoint route or sending several workers at several building jobs. Implementing building placeholders when a building order is given allows one to plan ahead base construction at moment when that it is convenient and minimizing hassle there too.

**Criticisms:**
Now I've had a bit of discussion about this on the OpenRA discord, and several valid criticisms have popped up, chief amongst which was along these lines:

To quote:
“The game is interesting exactly due to this, micromanagement, macro, multitasking. And the skill level depends on whether you can keep up with all that or not, so if you remove something you practically remove a part of skillbase, and the game loses dimensions. 
Problem with automation is that there are either too many different things to be automated, and it's hard to design UI properly (we already have a ton of modifiers on prod queue for example). Or you just automate it one way, everyone just uses it, and we get back to the problem of removing dimensions. “

Which is completely right of course, it is a skill to keep your building queue rolling through the chaotic and intense, and even the best players occasionally mess up with a front line tech center in the heat of the moment, or neglecting to place the next structure for half a minute. Automate building placement, and such occurrences will diminish greatly. (Although of course the queue can run out of things to build unnoticed more easily too now)

However, the same argument could be made against queuing in general. Yet ordering units one by one would be considered too distracting by most – and distracting from what then? The more time and focus must be spend on managing your queues, the less remains for managing the units on the battlefield. 

Generic “automation is bad” principles must still be weighted against the specific case proposed. If one wins by his opponent forgetting to keep his building queue rolling, this feels sort of cheap, if one loses by this it feels stupid. The ability to shoot oneself in the foot provides little venue for positive gameplay. It is not really something one can “play at” to get ones opponent to do, aside from giving them stress in general. But putting the pressure on has plenty of other ways to provide cracks.

There was also concern about base pushing becoming more viable, but I doubt one can reliably trust a ghost's location to remain unoccupied in such a scenario, and adjacency might very well be lost. For static defenses it might be a problem indeed, but if this turns out to be the case, the feature might just as well be disabled for the static defenses tab. 

The point was also raised that players might not know anymore how to manually place a building once they have gotten reliant on ghosts, but I doubt this will be something that's not intuitive, when building placement defaults to manual placement one sees the structure under the cursor after all.

**Conclusion:**

Implementing building ghosts to easier maintain the continued production of buildings would be a quality of life improvement, that's in line with the original design choices, improves the flow of the game, and removes scope for an unnecessary negative reward, at the cost of making the game slightly less skill intensive on the multitasking front.

**Comment by matjaeck** [comment]
> quality of life improvement, that's in line with the original design choices

Interesting idea, however to me this sounds like a complete new game. I wouldn't support this for the shipped mods. This is not just a multiplier but a whole new dimension of the game.

**Comment by Smittytron** [comment]
Closing this as wontfix after discussion in discord.

> pchote: again, could be done already with existing plumbing by a mod that wants it
> pchote: but not suitable for upstream mods
